lib/spack/llnl/util/filesystem.py

```diff
@@ -974,6 +974,49 @@ def remove_linked_tree(path):
             shutil.rmtree(path, True)
 
 
+@contextmanager
+def safe_remove(*files_or_dirs):
+    """Context manager to remove the files passed as input, but restore
+    them in case any exception is raised in the context block.
+
+    Args:
+        *files_or_dirs: glob expressions for files or directories
+            to be removed
+
+    Returns:
+        Dictionary that maps deleted files to their temporary copy
+        within the context block.
+    """
+    # Find all the files or directories that match
+    glob_matches = [glob.glob(x) for x in files_or_dirs]
+
+    # Copy them in a temporary location
+    removed, dst_root = {}, tempfile.mkdtemp()
+
+    try:
+        for id, file_or_dir in enumerate(itertools.chain(*glob_matches)):
+            src = os.path.abspath(file_or_dir)
+            # The glob expression at the top ensures that the file/dir exists
+            # at the time we enter the loop. Double check here since it might
+            # happen that a previous iteration of the loop already removed it.
+            # This is the case, for instance, if we remove the directory
+            # "/foo/bar" before the file "/foo/bar/baz.yaml".
+            if not os.path.exists(src):
+                continue
```

Comment:
This handles the first part where you have a removal order like `[x/, x/y]`. However, if you have a removal order like `[x/y, x/]`, then in the `except:` logic, if you try to add `x/y` back before `x/` you could encounter an error.

Sorting (considering shorter paths first) would resolve this (but you'd have to take the abspath of everything first).

Reply:
Done in 8982ee4